### PR TITLE
Germline mutation

### PR DESCRIPTION
**Pull request by rnugraha**
This PR has dependency on https://github.com/cBioPortal/cbioportal-frontend/pull/913

- Add germline mutation genetic rules
- Add background for rule legends

## Before 
schermafbeelding 2018-02-08 om 13 00 25
## After
after1

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["env"]
```
Comment (adamabeshouse) [inline .babelrc]: should this file be here?

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,7431 @@
+{
```
Comment (adamabeshouse) [inline package-lock.json]: do we need this file?

File: package.json [hunk 3]
```diff
@@ -38,6 +38,7 @@
   },
   "homepage": "https://github.com/cBioPortal/oncoprintjs",
   "devDependencies": {
+    "babel-preset-env": "^1.6.1",
```
Comment (adamabeshouse) [inline package.json]: needed?

File: rules/geneticrules.js [hunk 4]
```diff
@@ -9,353 +9,370 @@ var MUT_COLOR_TRUNC = '#000000';
...
-			'y': '33.33%',
-			'width': '100%',
-			'height': '33.33%',
-			'z': 6,
-		    }],
-		legend_label: 'Missense Mutation',
-	    },
-	}
+        'disp_mut': {
+            'promoter,promoter_rec,promoter_germ,promoter_rec_germ': {
```
Comment (adamabeshouse) [inline rules/geneticrules.js]: confused - so you both denote germline with the `disp_mut` value and the `disp_germline` value? why is that necessary?

File: src/js/oncoprintshapetosvg.js [hunk 5]
```diff
@@ -24,20 +24,29 @@ var extractColor = function(str) {
...
-	x: parseFloat(params.x) + offset_x,
-	y: parseFloat(params.y) + offset_y,
-	stroke: stroke_color.rgb,
-	'stroke-opacity': stroke_color.opacity,
-	'stroke-width': params['stroke-width'],
-	fill: fill_color.rgb,
-	'fill-opacity': fill_color.opacity
-    });
+    var el = document.createElementNS('http://www.w3.org/2000/svg', 'g');
+    el.appendChild(makeSVGRectBackground()); // add grey background
```
Comment (adamabeshouse) [inline src/js/oncoprintshapetosvg.js]: this feels hacky to me - to generate the rectangle background we should attach an extra rectangle Shape model to be rendered, instead of corrupting rectangleToSVG for a use-specific purpose

Comment (adamabeshouse) [reply src/js/oncoprintshapetosvg.js]: instead we should add something to `Rule` like `baseLegendShapes` or something. Just as you can see the use of the member `target` here: https://github.com/cBioPortal/oncoprintjs/blob/master/src/js/oncoprintlegendrenderer.js#L122 , which is something attached to the rule for the sole purpose of legend rendering

Comment (rnugraha) [reply src/js/oncoprintshapetosvg.js]: Which target in the code are you reffering to?

Comment (adamabeshouse) [reply src/js/oncoprintshapetosvg.js]: oops, looks like the file line number changed - now its here: https://github.com/cBioPortal/oncoprintjs/blob/master/src/js/oncoprintlegendrenderer.js#L156

`target` is a member of `rule.legend_config` that, if the rule is so configured, will be the data that the rule is applied to to generate the shape in the legend. 

It has basically nothing to do with this task except that its something attached to the rule thats used for the legend. So my idea is to add something like `base_shapes` to `legend_config`, and we can specify the gray rectangles there.

File: package-lock.json (hunk 2 above)
Comment (rnugraha) [reply package-lock.json]: Yes. It's advisable to always commit lock file to the source control.

File: rules/geneticrules.js (hunk 4 above)
Comment (rnugraha) [reply rules/geneticrules.js]: This to make a room of a possibility where germline mutations (all with '_germ' suffix) are displayed in different color or  different order.

Comment (adamabeshouse) [reply rules/geneticrules.js]: I see, that make sense. I'm concerned about how complicated this makes it now, with germ, rec, rec_germ. Maybe as long as we don't need separate graphics, we can just omit this - make disp_germ render as the small white line, and we can handle the different order inside the comparator function, which would have access to disp_mut and disp_germ

Comment (rnugraha) [reply rules/geneticrules.js]: I can't think of any better solution for this for now especially for dealing with the ordering and sorting in the cbioportal-frontend part

Comment (adamabeshouse) [reply rules/geneticrules.js]: what do you mean? we can add logic to check for disp_germline around oncoprint/SortUtils.ts:62

Comment (rnugraha) [reply rules/geneticrules.js]: There is also prioritization logic. https://github.com/thehyve/cbioportal-frontend/blob/3ab38b7188daab3034ef3efe047224cfd0c2775d/src/shared/components/oncoprint/DataUtils.ts#L33
Without the suffix, it's more challenging to prioritize the rendering.

Comment (adamabeshouse) [reply rules/geneticrules.js]: I just really don't like redundancy.. it makes maintenance more confusing and error-prone. In this case I think we could replace the `priority` map argument with a function that maps a datum to its order value, and that function will make use of the `disp_mut` value map, and the `disp_germline` value.

Comment (Sjoerd-van-Hagen) [reply rules/geneticrules.js]: @adamabeshouse is that the only thing blocking the merging of this PR and PR #913?

Comment (adamabeshouse) [reply rules/geneticrules.js]: @Sjoerd-van-Hagen yes I suppose so

File: src/js/oncoprintlegendrenderer.js [hunk 6]
```diff
@@ -148,14 +148,17 @@ var OncoprintLegendView = (function() {
...
     };
-    
+
     var ruleToSVGGroup = function(rule, view, model, target_svg, target_defs) {
 	var root = svgfactory.group(0,0);
 	var config = rule.getLegendConfig();
 	if (config.type === 'rule') {
 	    var concrete_shapes = rule.apply(config.target, model.getCellWidth(true), view.base_height);
+	    // generate backgrounds
+	    root.appendChild(svgfactory.bgrect(6, 20, 'rgb(190,190,190)'));
```
Comment (adamabeshouse) [inline src/js/oncoprintlegendrenderer.js]: confused.. this is not what we discussed

Comment (Sjoerd-van-Hagen) [reply src/js/oncoprintlegendrenderer.js]: You mean adding the gray backgrounds? I think it was discussed somewhere else and people agreed it was an improvement.

Comment (adamabeshouse) [reply src/js/oncoprintlegendrenderer.js]: not every rule will have a gray background, this is putting a very specific use case into a very general piece of code. what I had discussed with @rnugraha in the comment thread starting here https://github.com/cBioPortal/oncoprintjs/pull/55#discussion_r167663477 is adding a member to `Rule` that we can use for this purpose

File: rules/geneticrules.js (hunk 4 above)
Comment (rnugraha) [reply rules/geneticrules.js]: @adamabeshouse okay so I've changed disp_germline to isGermline and this flag is triggered by https://github.com/cBioPortal/cbioportal-frontend/pull/913/files#diff-bf5c432c23e8bb3755ca9a6b9da2af78R168

File: src/js/oncoprintlegendrenderer.js [hunk 7]
```diff
@@ -154,9 +154,11 @@ var OncoprintLegendView = (function() {
 	var config = rule.getLegendConfig();
 	if (config.type === 'rule') {
 	    var concrete_shapes = rule.apply(config.target, model.getCellWidth(true), view.base_height);
-	    // generate backgrounds
-	    root.appendChild(svgfactory.bgrect(6, 20, 'rgb(190,190,190)'));
-	    // generate shapes
+            if (config.baseLegendShapes) {
+                // generate backgrounds
+                root.appendChild(svgfactory.bgrect(6, 20, 'rgb(190,190,190)'));
```
Comment (adamabeshouse) [inline src/js/oncoprintlegendrenderer.js]: there was a misunderstanding - the idea is that in the rule we'd specify the particular baseLegendShapes, so baseLegendShapes is of type Shape[] not boolean

Comment (adamabeshouse) [reply src/js/oncoprintlegendrenderer.js]: if we're going to do it this way, which I think actually does make sense, we should change the name of baseLegendShapes to baseGrayRect or something

Comment (rnugraha) [reply src/js/oncoprintlegendrenderer.js]: will go for the 2nd option

Comment (adamabeshouse) [reply src/js/oncoprintlegendrenderer.js]: btw, please use `svgfactory.rect`, and use `model.getCellWidth(true)` and `view.base_height` instead of hardcoded 6,20

Comment (rnugraha) [reply src/js/oncoprintlegendrenderer.js]: why `rect' not `bgrect'?

Comment (adamabeshouse) [reply src/js/oncoprintlegendrenderer.js]: bgrect is just semantically for the white background. I mean its basically the same, just semantically thats not what it's meant for so I think it's cleaner to stick with rect.

**Comment by adamabeshouse** [comment]
looks good @rnugraha can you please squash to one commit?

**Comment by rnugraha** [comment]
I messed up the git history, I'll create new PR
